Fix envelope provider DEK decrypting after previous scheme

exosty opened the pull request:
DEK decrypting was failing if you tried to decrypt ciphertext which was previously encrypted with a different scheme without using envelope/DEK approach at all

### Detail

If your data is encrypted with non-envelope provider (for example with `DerivedSecretKeyProvider`) and now you want to change it to `EnvelopeEncryptionKeyProvider` using _previous scheme approach_:
  ```
    encrypts :phone, key_provider: ActiveRecord::Encryption::EnvelopeEncryptionKeyProvider.new,
                     previous: { key_provider: ActiveRecord::Encryption.key_provider }
  ```

you won't succeed because `EnvelopeEncryptionKeyProvider` expects DEK to be always present and decryption fails because of that:
```
NoMethodError: undefined method `payload' for nil:NilClass

          encrypted_data = encrypted_message.payload
                                            ^^^^^^^^
from ... activerecord-7.0.3.1/lib/active_record/encryption/cipher/aes256_gcm.rb:57:in `decrypt'
```


### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

exosty commented:
@jonathanhefner Hey. Can you review it please?

exosty commented:
@jonathanhefner ping 🙂  
would be nice either to have it either merged or cancelled

zzak commented:
@exosty Thanks for your PR, but please don't ping contributors. Someone will review it when they have time, if you want to get feedback we recommend posting it on the discord. :bow:

rafaelfranca commented on activerecord/test/cases/encryption/envelope_encryption_key_provider_test.rb:
```diff
@@ -29,6 +29,15 @@ class ActiveRecord::Encryption::EnvelopeEncryptionKeyProviderTest < ActiveRecord
     assert_equal key.secret, @key_provider.decryption_keys(encrypted_message).first.secret
   end
 
+  test "decryption_keys raises Errors::Decryption if encrypted message has no DEK" do
```
I don't think this tests demonstrate the problem you are trying to fix. Can you write a test that assets that the message will be decrypted by the previous provider automatically?

rafaelfranca replied:
Just to make clear, keep this test, but let's add a new test to make sure that raising `Errors::Decryption` actually allows the message to be decrypted.

exosty replied:
@rafaelfranca hey! Thanks for the review!
Did as you suggested - can you take a look please?

skipkayhil commented:
hey @exosty, can you rebase on `main`? I believe a lot of the CI failures you are seeing have been fixed and your branch just needs to include them

exosty committed:
Fix envelope provider DEK decrypting after previous scheme

DEK decrypting was failing if you tried to decrypt ciphertext which was previously encrypted with a different scheme without using envelope/DEK approach at all
